DOC: Move CUDA user guide from cuda-feedstock to conda-forge docs (#2702)
conda-forge/conda-forge.github.io: docs/maintainer/knowledge_base.mdx

for instructions on how to install the CUDA driver
([Linux](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html),
[Windows](https://docs.nvidia.com/cuda/cuda-installation-guide-microsoft-windows/index.html)).

:::tip

Conda CUDA packages are installable even if no CUDA driver or NVIDIA GPU is available, but
they will not run properly. See [Common problems and known
issues](#common-problems-and-known-issues) for why this is the case.

:::

#### Installing CUDA

##### Basic Installation

The easiest one-step solution to install the full CUDA Toolkit is to install the [`cuda`
metapackage](https://github.com/conda-forge/cuda-feedstock/) with this command:

```
conda install -c conda-forge cuda cuda-version=12
```

Let's break down this command. We are requesting the installation of two metapackages:
`cuda` and `cuda-version`.

The `cuda` metapackage pulls in all the components of the CUDA Toolkit (CTK) and is roughly
equivalent to installing the CUDA Toolkit with traditional system package managers like
`apt` or `yum` on Linux. Similarly to such package managers, the individual components may
also be installed independently.

The [`cuda-version`

> metapackage](https://github.com/conda-forge/cuda-version-feedstock/blob/main/recipe/README.md)

Reviewer: ```suggestion
The [`cuda-version` metapackage](
https://github.com/conda-forge/cuda-version-feedstock/blob/main/recipe/README.md )
```

is used to select the version of CUDA to install. This metapackage is important because
individual components of the CTK are typically versioned independently (the current versions
may be found in the [release

> notes](https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html)). The

Reviewer: ```suggestion
may be found in the [release notes](
https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html )). The
```

`cuda-version` metapackage provides a standard way to install the version of a specific CUDA
component corresponding to a given version of the CTK. This way, you never have to specify a
particular version of any CTK package; you just specify the `cuda-version` that you want,
then list packages you want installed and conda will take care of finding the right versions
for you. The above command will install all components of CUDA from the latest patch release

> of the CUDA 12.x series.

Reviewer: ```suggestion
component corresponding to a given version of the CTK. This way, one never has to specify a
particular version of any CTK package; one just specifies the `cuda-version` desired,
then list packages needed and conda will take care of finding the right versions.
The above command will install all components of CUDA from the latest patch release
of the CUDA 12.x series.
```

:::warning

The following CUDA components are not available as conda packages:

- Driver libraries, such as (but not limited to)
  - `libcuda`
  - `libnvidia-ml` library
  - GPUDirect Storage (GDS)
- Documentation and Samples
- Fabric Manager
- IMEX
- NVSwitch Configuration and Query (NSCQ)
- Nsight Systems

:::

##### Installing individual components of the CUDA Toolkit

Rather than installing all of CUDA at once, users may instead install just the packages that
they need. For example, to install just `libcublas` and `libcusparse` one may run:

```
conda install -c conda-forge libcublas libcusparse cuda-version=<CUDA version>
```

The best way to get a current listing is to run:

```
conda install --dry-run -c conda-forge cuda cuda-version=<CUDA version>
```

For a complete listing of the packages that were originally created, see [this
issue](https://github.com/conda-forge/staged-recipes/issues/21382).

##### Metapackages

For convenience, a number of additional metapackages are available:

- `cuda-runtime`: All CUDA runtime libraries needed to run a CUDA application

> - `cuda-libraries`: All libraries required to run a CUDA application requiring libraries beyond the CUDA runtime (such as the CUDA math libraries) as well as packages needed to perform JIT compilation

Reviewer: So [`cuda-runtime` just pulls in `cuda-libraries`]( https://github.com/conda-forge/cuda-runtime-feedstock/blob/060a7f720974b33e5fbbd88e1c2c121e50dc9773/recipe/meta.yaml#L19 ). So maybe we can drop `cuda-libraries` or include it as either/or

```suggestion
- `cuda-runtime` / `cuda-libraries`: All CUDA runtime libraries needed to run a CUDA application
```

> - `cuda-visual-tools`: GUIs for visualizing and profiling such as Nsight Compute

Reviewer: Needs a Trademark

```suggestion
- `cuda-visual-tools`: GUIs for visualizing and profiling such as Nsight™ Compute
```

- `cuda-command-line-tools`: Command line tools for analyzing and profiling such as cupti, cuda-gdb, and Compute Sanitizer
- `cuda-tools`: All tools for analyzing and profiling, both GUI (includes cuda-visual-tools) and CLI (includes cuda-command-line-tools)

> ##### Selecting ARM Variant

Reviewer: Maybe we should clarify this is related to a particular device

```suggestion
##### Selecting ARM Device Variant
```

Many of NVIDIA's CUDA packages are built for both SBSA and Tegra devices. For CUDA 12.9,
Tegra devices are a separate ARM variant. To install packages built for CUDA 12.9 Tegra, add
the `arm-variant=*=tegra` package to your environment. For CUDA 13.0 and later, the
arm-variant package is no longer required. Tegra devices are no longer a separate
architecture from SBSA.

:::info

Support for Tegra builds on conda-forge is only available for CUDA 12.9 and later. This
means that only Orin (sm_87) and later devices are supported.

:::

#### CUDA C++ Core Libraries (CCCL)

> CCCL is a special case among CUDA packages. Due to 1) being header-only, 2) fast-moving, and 3) independently-evolving, consumers may want a different (newer) version of CCCL than the

Reviewer: ```suggestion
CCCL is a special case among CUDA packages. Due to 1) being header-only, 2) fast-moving, and 3) independently-evolving,
consumers may want a different (newer) version of CCCL than the
```

one corresponding to their CTK version. Instructions on how to install a suitable CCCL
package from conda can be found
[in the CCCL README](https://github.com/NVIDIA/cccl/?tab=readme-ov-file#conda).
See [this issue](https://github.com/conda-forge/cuda-cccl-impl-feedstock/issues/2)
for more information on the history of these packages.

#### Package Naming Conventions

CUDA libraries are split into separate packages with standard naming:

- The base package (e.g., `libcublas`) provides runtime shared libraries.
- The `*-dev` package includes headers, symlinks, and CMake files for compiling code that dynamically links the library.

> - The `*-static` package includes the static library for static linking, and usually requires `*-dev` to be present for build files, though this isn't enforced.

Reviewer: ```suggestion
- The `*-static` package includes the static library for static linking
  For build files, users will also needed to install the `*-dev` packages.
```

on improving the quality of these installations, no corresponding updates were provided for
packages on the `nvidia` channel during this time period. While the `conda-forge` and
`nvidia` channel package lists were the same (i.e. the same packages existed in both places
with the same core contents like libraries and headers), the `nvidia` channel did not
include many of the incremental fixes made on `conda-forge` to improve things like symlinks,
static library handling, proper package constraints, etc. As a result, `nvidia` and
`conda-forge` CUDA packages remained incompatible from CUDA 12.0-12.4.

</details>
<details>
<summary>CUDA 12.5+</summary>

With CUDA 12.5, the `nvidia` channel was fully aligned with `conda-forge`. Packages on both
channels are identical, ensuring safe coexistence of the two channels within the same conda
environment.

Additionally [Anaconda has incorporated compatible CUDA 12 packages in its `defaults` channel]( https://www.anaconda.com/blog/anaconda-accelerates-ai-development-and-deployment-with-nvidia-cuda-toolkit).

Going forward, CUDA packages on the `conda-forge` and `nvidia` channels should be expected
to remain compatible.

> </details>

Reviewer: Will want to note that there are no further updates planned to the `nvidia` channel's CTK after 13.3.

```suggestion
</details>

<summary>CUDA 13.4+</summary>

CUDA 13.3, is the last CTK release to the `nvidia` channel.

Going forward, CUDA packages should be gotten from `conda-forge`.
</details>
```
